### ISSUE TITLE
ValidLogicalOperatorsSniff

### ISSUE THREAD

**Issue by vess** [photodude/coding-standards#10]
Should rewrite all `and` and `or` to `&&` and `||` used in if/elseif statements but not on things like `dothis() or die;` what fails with an error right now.

**Comment by photodude** [comment]
Looking over the code I think we can add another token check for if the next token is `exit()` or `die()` to solve part of this issue. Since checking for `T_EXIT` being the next token should cover situations for `defined('_JEXEC') or die;`

Unfortunately, that kind of fix won't apply to things like `JSession::checkToken() or jexit(JText::_('JINVALID_TOKEN'));` since `jexit()` is not tokenized and wouldn't be covered by the `T_EXIT` token.

**Comment by vess** [comment]
> since jexit() is not tokenized and wouldn't be covered by the T_EXIT token.

Right.
My head isn't really in the coding mode right now, since he was asleep for nearly 5 months, but I guess there are only 2 options:
1. checking if we are within an if/elseif
or 2. check the next token (excluding the whitespace) for token code T_EXIT and token content 'jexit...'

**Comment by vess** [comment]
Ok you're already testing with token content as I can see now when looking at the PR

**Comment by photodude** [comment]
Off the top of my head when I was thinking about the issue I thought just checking against the next token would be easier than checking against all the different logical opperators.

Following your suggestion, and looking at the PHPCS "examples" the resulting fix was 
`$nextToken = $phpcsFile->findNext(T_WHITESPACE, ($stackPtr + 1), null, true);`

so the `defined('_JEXEC') or die;` issue is now fixed.

I'm leaving the issue open since `JSession::checkToken() or jexit(JText::_('JINVALID_TOKEN'));` is still an issue.

**Comment by vess** [comment]
Great job @photodude now it is working like it should.
For the special Joomla case can you try to add another if statement after the one you added pls.

// Special Joomla case.
if ($tokens[$nextToken]['content'] === 'jexit(JText::_('JINVALID_TOKEN')')
{
return;
}

but I'm really not sure if the content is the complete function call, maybe only the function name

**Comment by photodude** [comment]
Ideally, it needs to be checking for just the method name `jexit()` as the method might have a different exit message depending on the context it's used in.

I've attempted the following, unfortunately it didn't work. (still trying to understand the `code` vs `content` differences)

if ($tokens[$nextToken]['content'] == 'jexit')
{
    return;
}

**Comment by vess** [comment]
if ($tokens[$nextToken]['content'] === 'jexit')
{
return;
}

that does however work with my test file (phpcs 2.3.4)

**Comment by vess** [comment]
Ok be prepared for all the special cases we're dealing with @photodude hehe

**/administrator/components/com_finder/controllers/indexer.json.php**
JSession::checkToken('request') or $this->sendResponse(new Exception(JText::_('JINVALID_TOKEN'), 403));

**/administrator/components/com_finder/controllers/indexer.json.php**
lot of 'or' usage like:
self::_replace($word, 'sses', 'ss')
or self::_replace($word, 'ies', 'i')
or self::_replace($word, 'ss', 'ss')
or self::_replace($word, 's', '');

lot of 'and' usage like:
if (!self::_replace($word, 'at', 'ate') and !self::_replace($word, 'bl', 'ble') and !self::_replace($word, 'iz', 'ize'))
{
    // Double consonant ending
    if (self::_doubleConsonant($word) and substr($word, -2) != 'll' and substr($word, -2) != 'ss' and substr($word, -2) != 'zz')
    {
        $word = substr($word, 0, -1);
    }
    elseif (self::_m($word) == 1 and self::_cvc($word))
    {
        $word .= 'e';
    }
}

**/administrator/components/com_installer/controllers/update.php**
(JSession::checkToken() or JSession::checkToken('get')) or jexit(JText::_('JINVALID_TOKEN'));

**/administrator/components/com_joomlaupdate/restore.php**
defined('DS') or define('DS', DIRECTORY_SEPARATOR);

**/administrator/components/com_media/media.php**
if (!$user->authorise('core.manage', 'com_media') && (!$asset or (!$user->authorise('core.edit', $asset)

**/components/com_contact/router.php**
l61: if (isset($view) && ($mView == $view) and (isset($query['id'])) and ($mId == (int) $query['id']))
l69: if (isset($view) and ($view == 'category' or $view == 'contact'))

**/components/com_newsfeeds/router.php**
l61: if (isset($query['view']) && ($mView == $query['view']) and (isset($query['id'])) and ($mId == (int) $query['id']))
l70: if (isset($view) and ($view == 'category' or $view == 'newsfeed'))

**/components/com_newsfeeds/views/newsfeed/view.html.php**
l152: if (!in_array($item->access, $levels) or ((in_array($item->access, $levels) and (!in_array($item->category_access, $levels)))))

**/installation/controller/database.php
/installation/controller/detectftproot.php
/installation/controller/ftp.php
/installation/controller/install/config.php
/installation/controller/install/database.php
/installation/controller/install/database_backup.php
/installation/controller/install/sample.php
/installation/controller/preinstall.php
/installation/controller/removefolder.php
/installation/controller/setlanguage.php
/installation/controller/site.php
/installation/controller/summary.php
/installation/controller/verifyftpsettings.php**
l33: JSession::checkToken() or $app->sendJsonResponse(new Exception(JText::_('JINVALID_TOKEN'), 403));

**/installation/controller/install/languages.php
/installation/controller/setdefaultlanguage.php**
l47: JSession::checkToken() or $app->sendJsonResponse(new Exception(JText::_('JINVALID_TOKEN'), 403));

**Comment by vess** [comment]
In reality there not that much "special cases" here
1. defined('DS') or define('DS', DIRECTORY_SEPARATOR);  <-- if we don't consider removing it at all
2. JSession::checkToken('request') or $this->sendResponse(new Exception(JText::_('JINVALID_TOKEN'), 403));
and
JSession::checkToken() or $app->sendJsonResponse(new Exception(JText::_('JINVALID_TOKEN'), 403));

all other ones can be replaced by the auto fixer I would say

**Comment by photodude** [comment]
Some of those "special cases" should be handled, others need to use the correct logical operators especially those in control statements. 
'jexit' is still failing
'or define' should probably be handed (or throw a notice that it's something needing  to fixed as it's related to a Joomla version deprecated code item)

I'll try to look at things again next weekend, I need to focus on finalizing a couple of others projects this week.

**Comment by vess** [comment]
Pls give me something to look where 'jexit' is failing, since it is working here on my laptop, which version of phpcs are you using?
And in the travis log you posted there is no 'jexit' error, only the ones I posted above

**Comment by photodude** [comment]
The [one I was noticing seems to be "more complex"](https://github.com/joomla/joomla-cms/blob/staging/administrator/components/com_installer/controllers/update.php#L83)

That one might be failing on the `JSession::checkToken()` rather than on the `jexit()`

I'm running off the master branch (2.6.0 alpha) of PHPCS since 2.6.0 has yet to be released. I'm running off that because there are some errors in 2.5.1 related to the JS tokenizer which will crash phpcs on some of the library code in the CMS.

**Comment by vess** [comment]
Ok all is fine, that one is also a special, I will include it and will post the updated if statement that should work with all the cases.
but for that one it would really be nicer to write it
(JSession::checkToken() || JSession::checkToken('get')) or jexit(JText::_('JINVALID_TOKEN'));
in the core.

**Comment by photodude** [comment]
Sounds good, I just added a check for the `JSession::checkToken()` case, It seems to be working. 

The only ones left are the two with variables starting `or $this->sendResponse()` and `or $app->sendJsonResponse()`

I'm not sure about `or define()` leaning towards it's a workaround for old code and should give a different notice about it's use.

**Comment by vess** [comment]
Giving a different notice for the define would be to much, let us just include it and when that statement is gone from the core we can remove it.
I do a test for the JSession::checkToken() case in a few minutes, but I would say we should be more granular with these ones since they maybe will occur in some if calls as well.

**Comment by photodude** [comment]
Maybe that's a reason to go back to the other concept of only enforcing "Valid Logical Operators" for control statements. That might catch all the other exceptions we are trying to handle now.

**Comment by vess** [comment]
Well maybe that would do.
For this right now, that should do the trick:

```
    $next = $phpcsFile->findNext(T_WHITESPACE, ($stackPtr + 1), null, true);

    if ($tokens[$next]['code'] === T_EXIT)
    {
        // Put in an exception for things like `or die;` and `or exit;`
        return;
    }

    // Special Joomla! cases.
    if ($tokens[$next]['content'] === 'jexit'
            || $tokens[$next]['content'] === 'JSession' && $tokens[($next + 10)]['content'] === 'jexit'
            || $tokens[$next]['content'] === 'define'
            || $tokens[($next + 2)]['content'] === 'sendResponse'
            || $tokens[($next + 2)]['content'] === 'sendJsonResponse')
    {
        // Put in an exception for things like `or jexit()`
        return;
    }
```

**Comment by vess** [comment]
If we haven't overseen a special one the auto fixer should only change the ones we want it to, fingers crossed^^
But going back to enforcing only control statements might be better for future weird special cases...
If you ask me, those "special" ones should be done with proper if statements, but that's only my opinion.

**Comment by photodude** [comment]
Thanks that change seems to work great. I'm considering this issue solved with 8fffca9
